### PR TITLE
feat(ivy): support host attribute and property bindings

### PR DESCRIPTION
**Pull request by chuckjaz**
## PR Type
What kind of change does this PR introduce?

```
[x] Feature
```

## What is the current behavior?

Ivy compiler does not support host bindings.

## What is the new behavior?

Ivy compiler supports host bindings.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 1]
```diff
@@ -355,16 +358,17 @@ class TemplateDefinitionBuilder implements TemplateAstVisitor, LocalResolver {
...
 
         // (r3.qR(tmp = r3.ɵld(0)) && (ctx.someDir = tmp));
         const temporary = this.temp();
         const getQueryList = o.importExpr(R3.load).callFn([o.literal(querySlot)]);
         const refresh = o.importExpr(R3.queryRefresh).callFn([temporary.set(getQueryList)]);
-        const updateDirective = o.variable(CONTEXT_NAME).prop(query.propertyName).set(
-          query.first ? temporary.key(o.literal(0)) : temporary);
+        const updateDirective = o.variable(CONTEXT_NAME)
+                                    .prop(query.propertyName)
+                                    .set(query.first ? temporary.key(o.literal(0)) : temporary);
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: same comment as previous PR

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 2]
```diff
@@ -749,61 +752,135 @@ export function createFactory(
...
-export function createHostBindingsFunction(type: CompileTypeMetadata, outputCtx: OutputContext,
-  queries: CompileQueryMetadata[]): o.Expression | null {
+type HostBindings = {
+  [key: string]: string
+};
 
+function createHostAttributesArray(
+    directiveMetadata: CompileDirectiveMetadata, outputCtx: OutputContext): o.Expression|null {
+  const values: o.Expression[] = [];
+  const attributes = directiveMetadata.hostAttributes;
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: `Object.getOwnPropertyNames()`

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 1 above)
Comment (vicb) [reply packages/compiler/src/render3/r3_view_compiler.ts]: _because it is the same code :)_

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 3]
```diff
@@ -749,61 +752,135 @@ export function createFactory(
...
     const getDirectiveMemory = o.importExpr(R3.load).callFn([o.variable('dirIndex')]);
     // The query list is at the query index + 1 because the directive itself is in slot 0.
     const getQueryList = getDirectiveMemory.key(o.literal(index + 1));
     const assignToTemporary = temporary().set(getQueryList);
     const callQueryRefresh = o.importExpr(R3.queryRefresh).callFn([assignToTemporary]);
-    const updateDirective = getDirectiveMemory.key(o.literal(0, o.INFERRED_TYPE)).prop(query.propertyName).set(
-      query.first ? temporary().key(o.literal(0)) : temporary());
+    const updateDirective = getDirectiveMemory.key(o.literal(0, o.INFERRED_TYPE))
+                                .prop(query.propertyName)
+                                .set(query.first ? temporary().key(o.literal(0)) : temporary());
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: here also

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 4]
```diff
@@ -76,46 +85,44 @@ export function compileDirective(
 
 export function compileComponent(
     outputCtx: OutputContext, component: CompileDirectiveMetadata, pipes: CompilePipeSummary[],
-    template: TemplateAst[], reflector: CompileReflector) {
+    template: TemplateAst[], reflector: CompileReflector, bindingParser: BindingParser) {
   const definitionMapValues: {key: string, quoted: boolean, value: o.Expression}[] = [];
 
+  const field = (key: string, value: o.Expression | null) => {
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: Not sure if 2 is enough for a `fieldBuilderBuilder()` ? May be if you anticipate more usage

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 5]
```diff
@@ -808,21 +834,70 @@ export function createHostBindingsFunction(
...
+  if (eventBindings) {
+    for (const binding of eventBindings) {
+      const bindingExpr = convertActionBinding(
+          null, bindingContext, binding.handler, 'b', () => error('Unexpected interpolation'));
+      const bindingName = binding.name && sanitizeIdentifier(binding.name);
+      const typeName = identifierName(directiveMetadata.type);
+      const functionName =
+          typeName && bindingName ? `${typeName}_${bindingName}_HostBindingHandler` : null;
+      const handler = o.fn(
+          [new o.FnParam('event', o.DYNAMIC_TYPE)],
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: should 'b' and 'event' be consts ?

File: packages/compiler/src/template_parser/binding_parser.ts [hunk 6]
```diff
@@ -78,11 +77,19 @@ export class BindingParser {
...
+      return boundProps;
     }
     return null;
   }
 
+  createDirectiveHostPropertyAsts(
+      dirMeta: CompileDirectiveSummary, elementSelector: string,
+      sourceSpan: ParseSourceSpan): BoundElementPropertyAst[]|null {
+    const boundProps = this.createBoundHostProperties(dirMeta, sourceSpan);
+    return boundProps &&
```
Comment (vicb) [inline packages/compiler/src/template_parser/binding_parser.ts]: nit: `return boundProps ? map : null` (only style change)

File: packages/compiler/test/render3/mock_compile.ts [hunk 7]
```diff
@@ -243,5 +251,9 @@ export function compile(
       /* referenceFilter */ undefined,
       /* importFilter */ e => e.moduleName != null && e.moduleName.startsWith('/app'));
 
+  if (errors.length) {
+    throw new Error('Unexpected errors:' + errors.map(e => e.toString).join(', '));
```
Comment (vicb) [inline packages/compiler/test/render3/mock_compile.ts]: nit: is `map(e => e.toString)` required here ?

**Review by vicb — APPROVED** [review]
a few nits

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 4 above)
Comment (chuckjaz) [reply packages/compiler/src/render3/r3_view_compiler.ts]: Not really but it is more consistent with the one above.

File: packages/compiler/src/template_parser/binding_parser.ts (hunk 6 above)
Comment (chuckjaz) [reply packages/compiler/src/template_parser/binding_parser.ts]: I prefer the coalescing pattern. It is sad that JavaScript doesn't have a real one but for object references && works.

File: packages/compiler/test/render3/mock_compile.ts (hunk 7 above)
Comment (chuckjaz) [reply packages/compiler/test/render3/mock_compile.ts]: This doesn't emit the right result, the fixed, toString() call.

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 5 above)
Comment (chuckjaz) [reply packages/compiler/src/render3/r3_view_compiler.ts]: `event` is a parameter and 'b' is the context which is generated in the `convertActionBinding` which is not part of this change.

**Pull request closed without merging**